lib/auth/bot_test.go

```diff
@@ -97,3 +230,145 @@ func TestBotOnboardFeatureDisabled(t *testing.T) {
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "not licensed")
 }
+
+func renewBotCerts(
+	srv *TestTLSServer, tlsCert tls.Certificate, botUser string,
+	publicKey []byte, privateKey []byte,
+) (*Client, *proto.Certs, tls.Certificate, error) {
+	client := srv.NewClientWithCert(tlsCert)
+
+	certs, err := client.GenerateUserCerts(context.Background(), proto.UserCertsRequest{
+		PublicKey: publicKey,
+		Username:  botUser,
+		Expires:   time.Now().Add(1 * time.Hour),
+	})
+	if err != nil {
+		return nil, nil, tls.Certificate{}, trace.Wrap(err)
+	}
+
+	// Make sure to overwrite tlsCert with the new certs.
+	tlsCert, err = tls.X509KeyPair(certs.TLS, privateKey)
+	if err != nil {
+		return nil, nil, tls.Certificate{}, trace.Wrap(err)
+	}
+
+	return client, certs, tlsCert, nil
+}
+
+// TestRegisterBotCertificateGenerationCheck ensures bot cert generation checks
+// work in ordinary conditions, with several rapid renewals.
+func TestRegisterBotCertificateGenerationCheck(t *testing.T) {
+	t.Parallel()
+	srv := newTestTLSServer(t)
+
+	_, err := CreateRole(context.Background(), srv.Auth(), "example", types.RoleSpecV5{})
+	require.NoError(t, err)
+
+	// Create a new bot.
+	bot, err := srv.Auth().createBot(context.Background(), &proto.CreateBotRequest{
```

Comment:
The same here and in other tests. Can we create one `ctx` and pass it everywhere where is needed instead of creating a new context for each call?

Reply:
https://github.com/gravitational/teleport/pull/15117/commits/3064ed7c6b9558fe8d43c53900bd62582f610ca4